libs/ojsonnet/Eval_jsonnet.ml

```diff
@@ -670,6 +668,22 @@ and manifest_value (v : V.value_) : JSON.t =
                    let new_self = V.Object (_l, (_assertsTODO, fields), _r) in
                    let locals =
                      fld_value.env.locals
+                     (* We need to do these assignment on field access rather than on
+                        object creation, since when objects are merged, we need self to
+                        reference the new merged object rather than the original. Here's
+                        such an example:
+                        ({ name : self.y } + {y : 42})["name"]
+                        If we were to do the assignment of self before doing the field access
+                        we would have the following (incorrect) evaluation where o = { name : self.y }
+                        ({name : o.y} + {y : 42})["name"]
+                        o.y
+                        {name : self.y}.y
+                        Error no such field.
+                        However, if we only assign self on access, we get the following (correct) evaluation
+                        ({ name : self.y } + {y : 42})["name"]
+                        { name : self.y, y : 42 }["name"]
+                        {name: self.y, y : 42}[y]
+                        42 *)
                      |> Map_.add V.LSelf
                           { V.value = Val new_self; env = fld_value.env }
                      |> Map_.add V.LSuper
```

Comment:
Also is this correct? Seems like Super is always bound to an empty_obj. I know you have more PRs fixing + later, but maybe would be worth to add a TODO comment saying this handling of LSuper is incorrect and need to be fixed.

Reply:
Actually I think I understand. This is correct code probably. By default super is the empty obj, 
it's just in + that we do something special and bind it to the left object.